### PR TITLE
fix runtime-config issue  due to Initializers is enabled by kubeadm

### LINKED ISSUE #629: Kubeadm enables Initializer admission controller by default
When deploying a 1.9 cluster, kubeadm adds `Initializers` to `--admission-control` by default, but doesn't enable its corresponding API `admissionregistration.k8s.io/v1alpha1` in `--runtime-config`. See [docs](https://kubernetes.io/docs/admin/extensible-admission-controllers/#enable-initializers-alpha-feature)

This causes 404s to be spammed constantly in the request audit logs.

Also this is an alpha feature, shouldn't it be disabled by default?

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
`kubeadm` adds Initializers to `--admission-control` by default since v1.8, but doesn't enable its corresponding API `admissionregistration.k8s.io/v1alpha1` in `--runtime-config`.

Currently only webhook part API of `admissionregistration` is graduated to `v1beta1` by #56004. For `Initializers`, we still need to register `v1alpha1` to runtime config until it is promoted to `v1beta1`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#629

**Special notes for your reviewer**:
/assign @luxas
/area kubeadm
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews 

**Release note**:

```release-note
add "admissionregistration.k8s.io/v1alpha1" to the --runtime-config flag due to Initializers is enabled by kubeadm
```

### CODE REVIEW

**Comment by zouyee** [comment]
could not we config this value using apiServerExtraArgs?

**Comment by dixudx** [comment]
> could not we config this value using apiServerExtraArgs?

This value should be set by default, since `Initializers` has already been added into `admission-control` by default. Of course we can use `apiServerExtraArgs` to set this. But `"admissionregistration.k8s.io/v1alpha1"` and `Initializers` are coupled now (v1alpha1 is not registered by default), it is not right to split them separately and let users manually input this value.

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 1]
```diff
@@ -163,6 +163,8 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, k8sVersion *versio
 		"requestheader-allowed-names":        "front-proxy-client",
 		"proxy-client-cert-file":             filepath.Join(cfg.CertificatesDir, kubeadmconstants.FrontProxyClientCertName),
 		"proxy-client-key-file":              filepath.Join(cfg.CertificatesDir, kubeadmconstants.FrontProxyClientKeyName),
+		// add "admissionregistration.k8s.io/v1alpha1" to the --runtime-config flag due to Initializers is enabled
```
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: nit: replace `due to Initializers is enabled` with `because Initializers are enabled`

**Review by jamiehannaford — CHANGES_REQUESTED** [review]
just one nit. overall looks good, thanks for submitting

**Review by fabriziopandini — APPROVED** [review]
Thanks for fixing this!
/lgtm

**Comment by dixudx** [comment]
@fabriziopandini  @jamiehannaford Update the nit. PTAL. Thanks.

**Comment by xiangpengzhao** [comment]
@dixudx the admissionregistration API is beta and enabled by default in 1.9 (https://github.com/kubernetes/kubernetes/pull/56687), so we don't need to pass any admissionregistration API version to `--runtime-config`.

**Comment by dixudx** [comment]
> the admissionregistration API is beta and enabled by default in 1.9 (#56687), so we don't need to pass any admissionregistration API version to --runtime-config.

@xiangpengzhao Thanks for raising this question. Currently we only graduate `admissionregistration` to `v1beta1`, but the [informer](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/initialization/initialization.go#L100) still uses `v1alpha1`. This fix is only for `release-1.9`.

You're right. For `master` branch, we may use `v1beta1` instead, which is registered by default. Will submit another PR to master.

**Review by jamiehannaford — APPROVED** [review]
/lgtm

**Comment by dixudx** [comment]
Can anybody remove the label `do-not-merge/cherry-pick-not-approved`. This PR is targeted for master branch. Thanks.

**Comment by dixudx** [comment]
ping @luxas for approval. Thanks.

**Comment by dixudx** [comment]
Seems `pull-kubernetes-e2e-gke-gci` is not happy with this PR.

Rebased. PTAL.

**Comment by timothysc** [comment]
Umm do we want to support "Initializers?" or is is mutating web-hooks or bust? 

/cc @mattmoyer @ncdc

**Comment by jamiehannaford** [comment]
This should probably tie in to a wider policy of how kubeadm promotes unstable/alpha features. I think initializers are a useful enough feature to have, plus users still have the choice of electing to use them.

**Comment by timothysc** [comment]
@jamiehannaford ... Sure, but I think they are planned to be deprecated, but I want @kubernetes/sig-api-machinery-misc to chime in on the "official" stance on this feature.

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 2]
```diff
@@ -163,6 +163,8 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, k8sVersion *versio
 		"requestheader-allowed-names":        "front-proxy-client",
 		"proxy-client-cert-file":             filepath.Join(cfg.CertificatesDir, kubeadmconstants.FrontProxyClientCertName),
 		"proxy-client-key-file":              filepath.Join(cfg.CertificatesDir, kubeadmconstants.FrontProxyClientKeyName),
+		// add "admissionregistration.k8s.io/v1alpha1" to the --runtime-config flag because Initializers are enabled
+		"runtime-config": "admissionregistration.k8s.io/v1alpha1",
```
Comment (liggitt) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Definitely don't enable alpha versions by default.

**Comment by liggitt** [comment]
The alpha api version should not be enabled by default. 

It would be more appropriate to remove the initializers admission plugin from the default chain

/close

**Comment by jamiehannaford** [comment]
@dixudx Would you mind removing Initializers from the admission chain?

**Comment by dixudx** [comment]
@jamiehannaford Remove `Initializers` in PR #58428. PTAL. Thanks.

**Pull request closed without merging**